indexer-app/src/cmr/indexer/data/concepts/generic.clj

```diff
@@ -0,0 +1,120 @@
+   [cmr.common.mime-types :as mtype]
+   [cmr.common.util :as util]
+   [cmr.indexer.data.concept-parser :as concept-parser]
+   [cmr.indexer.data.concepts.generic-util :as gen-util]
+   [cmr.indexer.data.concepts.keyword-util :as keyword-util]
+   [cmr.indexer.data.elasticsearch :as esearch]))
+  
+(defmulti field->index
+  "Functions which convert a part of metadata to a name-value which can be added
+   to an index document. This defmulti is directed by looking for an :Indexer
+   value in settings and assuming :default if it is not set.
+   Usage:
+   settings - json configuration for one field
+   data - metadata document
+   (field->index
+            {:Name 'complex-field'
+             :Field '.l1'
+             :Indexer 'complex-field'
+             :Configuration {:sub-fields [:s2] :format '%s == %s'}}
+            {:l1 {:s1 'one' :s2 'two'}})"
+  (fn [settings _] (or (:Indexer settings) :default)))
+  
+(defmethod field->index "complex-field"
+  ;; This is an example of a complex indexer which takes a list of sub fields and
+  ;; combines them into one field"
+  [settings data]
+  (let [field-list (get settings :Field ".")
+        field-data (get-in data (gen-util/jq->list field-list keyword) {})
+        config (get settings :Configuration {})
+        sub-fields (get config :sub-fields {})
+        layout (get config :format "%s=%s")
+        field-name (util/safe-lowercase (:Name settings))
+        field-name-lower (str field-name"-lowercase")
+        field-value (reduce (fn [data-str, key-name]
+                              (str
+                               data-str
+                               (when-not (empty? data-str) ", ")
+                               (format layout key-name (get field-data (keyword key-name)))))
+                            ""
+                            sub-fields)
+        field-value-lower (util/safe-lowercase field-value)]
+    {(keyword field-name) field-value
+     (keyword field-name-lower) field-value-lower}))
+
+(defmethod field->index :default
+  ;; The default indexer which will map one metadata field to two indexes. One is
+  ;; with the literal case, another is all lower case"
+  [settings data]
+  (let [field-name (util/safe-lowercase (:Name settings))
+        field-name-lower (str field-name "-lowercase")
+        value (get-in data (gen-util/jq->list (:Field settings) keyword))
+        value-lower (util/safe-lowercase value)]
+    {(keyword field-name) value
+     (keyword field-name-lower) value-lower}))
+
+(defn- parsed-concept->elastic-doc-without-context
+  "Generate an all of the elastic document parts that do not require a context"
+  [concept parsed-concept]
+  (let [{:keys [concept-id revision-id deleted provider-id user-id
+                revision-date format-key extra-fields native-id]} concept
+        long-name (:LongName parsed-concept) ; should this exist as a required field
+        ;; TODO: Generic work: Need to remove this section 
```

Comment:
Please honor this TODO and remove the section or update the comment

Reply:
https://bugs.earthdata.nasa.gov/browse/CMR-8616